packages/ipa-certified-attributes-importer/src/index.ts

```diff
@@ -0,0 +1,467 @@
+  const platformAttributeIndex = new Set(
+    attributes
+      .filter((a) => a.kind === "Certified" && a.origin && a.code)
+      .map((a) => toKey({ origin: a.origin, code: a.code }))
+  );
+
+  const newAttributesIndex = new Set(
+    tenantUpsertData.flatMap((t) =>
+      t.attributes.map((a) => toKey({ origin: a.origin, code: a.code }))
+    )
+  );
+
+  return registryData.attributes.filter(
+    (a) =>
+      newAttributesIndex.has(toKey({ origin: a.origin, code: a.code })) &&
+      !platformAttributeIndex.has(toKey({ origin: a.origin, code: a.code }))
+  );
+}
+
+export async function getAttributesToAssign(
+  platformTenant: Tenant[],
+  platformAttributes: Attribute[],
+  tenantSeed: TenantSeed[]
+): Promise<tenantApi.InternalTenantSeed[]> {
+  const tenantIndex = new Map(
+    platformTenant.map((t) => [toKey(t.externalId), t])
+  );
+
+  const certifiedAttribute = new Map(
+    platformAttributes
+      .filter((a) => a.kind === "Certified" && a.origin && a.code)
+      .map((a) => [a.id, a])
+  );
+
+  return tenantSeed
+    .map((i) => {
+      const externalId = { origin: i.origin, value: i.originId };
+
+      const tenant = tenantIndex.get(toKey(externalId));
+
+      if (!tenant) {
+        return undefined;
+      }
+
+      const tenantCurrentAttribute = new Map(
+        tenant.attributes
+          .map((a) => {
+            const withRevocation = match(a)
+              .with({ type: "PersistentCertifiedAttribute" }, (certified) => ({
+                ...a,
+                revocationTimestamp: certified.revocationTimestamp,
+              }))
+              .otherwise((_) => undefined);
+
+            if (withRevocation) {
+              const attribute = certifiedAttribute.get(withRevocation.id);
+
+              if (attribute) {
+                return {
+                  ...attribute,
+                  revocationTimestamp: withRevocation.revocationTimestamp,
+                };
+              }
+            }
+
+            return undefined;
+          })
+          .filter((a) => a !== undefined)
+          .map((a) => [toKey({ origin: a?.origin, code: a?.code }), a])
+      );
+
+      return tenant
+        ? {
+            externalId,
+            name: tenant.name,
+            certifiedAttributes: i.attributes
+              .filter((a) => {
+                const attribute = tenantCurrentAttribute.get(
+                  toKey({
+                    origin: a.origin,
+                    code: a.code,
+                  })
+                );
+
+                if (!attribute) {
+                  return true;
+                }
+
+                return attribute.revocationTimestamp !== undefined;
+              })
+              .map((a) => ({
+                origin: a.origin,
+                code: a.code,
+              })),
+          }
+        : undefined;
+    })
+    .filter(
+      (t) => t !== undefined && t.certifiedAttributes.length > 0
+    ) as tenantApi.InternalTenantSeed[];
```

Comment:
Is this `as` needed?

Reply:
Yes, because in reality is still an `InternalTenantSeed | undefined`. Maybe I should write a type guard